crates/ironrdp-server/src/encoder/mod.rs

```diff
@@ -408,6 +448,101 @@ impl BitmapUpdateHandler for RemoteFxHandler {
     }
 }
 
+#[cfg(feature = "qoi")]
+#[derive(Clone, Debug)]
+struct QoiHandler {
+    codec_id: u8,
+}
+
+#[cfg(feature = "qoi")]
+impl QoiHandler {
+    fn new(codec_id: u8) -> Self {
+        Self { codec_id }
+    }
+}
+
+#[cfg(feature = "qoi")]
+impl BitmapUpdateHandler for QoiHandler {
+    fn handle(&mut self, bitmap: &BitmapUpdate) -> Result<UpdateFragmenter> {
+        let data = qoi_encode(bitmap)?;
+        set_surface(bitmap, self.codec_id, &data)
+    }
+}
+
+#[cfg(feature = "qoiz")]
+#[derive(Clone)]
+struct QoizHandler {
+    codec_id: u8,
+    zctxt: Arc<Mutex<zstd_safe::CCtx<'static>>>,
+}
+
+#[cfg(feature = "qoiz")]
+impl fmt::Debug for QoizHandler {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("QoizHandler").field("codec_id", &self.codec_id).finish()
+    }
+}
+
+#[cfg(feature = "qoiz")]
+impl QoizHandler {
+    fn new(codec_id: u8) -> Self {
+        let mut zctxt = zstd_safe::CCtx::default();
+
+        zctxt.set_parameter(zstd_safe::CParameter::CompressionLevel(3)).unwrap();
+        zctxt
+            .set_parameter(zstd_safe::CParameter::EnableLongDistanceMatching(true))
+            .unwrap();
+        let zctxt = Arc::new(Mutex::new(zctxt));
+
+        Self { codec_id, zctxt }
+    }
+}
+
+#[cfg(feature = "qoiz")]
+impl BitmapUpdateHandler for QoizHandler {
+    fn handle(&mut self, bitmap: &BitmapUpdate) -> Result<UpdateFragmenter> {
+        let qoi = qoi_encode(bitmap)?;
+        let mut inb = zstd_safe::InBuffer::around(&qoi);
+        let mut data = vec![0; qoi.len()];
+        let mut outb = zstd_safe::OutBuffer::around(data.as_mut_slice());
+
+        let mut zctxt = self.zctxt.lock().unwrap();
+        let res = zctxt
+            .compress_stream2(
+                &mut outb,
+                &mut inb,
+                zstd_safe::zstd_sys::ZSTD_EndDirective::ZSTD_e_flush,
+            )
+            .map_err(zstd_safe::get_error_name)
+            .unwrap();
```

Comment:
question: Why is it fine to call unwrap at this place?

Reply:
you are right, I looked over it as it doesn't happen in my tests. I'll fix it.